src/sparse.jl

```diff
@@ -0,0 +1,164 @@
+
+Use `ca` to construct a coloring vector `c` of length `size(M, 2)` such that if two columns `j1` and `j2` satisfy `c[j1] = c[j2]`, they do not share any nonzero coefficients in `M`.
+"""
+function column_coloring end
+
+"""
+    row_coloring(M::AbstractMatrix, ca::ColoringAlgorithm)::AbstractVector{<:Integer}
+
+Use `ca` to construct a coloring vector `c` of length `size(M, 1)` such that if two rows `i1` and `i2` satisfy `c[i1] = c[i2]`, they do not share any nonzero coefficients in `M`.
+"""
+function row_coloring end
+
+"""
+    NoColoringAlgorithm <: AbstractColoringAlgorithm
+
+Trivial coloring algorithm, which always returns a different color for each matrix column/row.
+
+# See also
+
+- [`AbstractColoringAlgorithm`](@ref)
+"""
+struct NoColoringAlgorithm <: AbstractColoringAlgorithm end
+
+column_coloring(M::AbstractMatrix, ::NoColoringAlgorithm) = 1:size(M, 2)
+row_coloring(M::AbstractMatrix, ::NoColoringAlgorithm) = 1:size(M, 1)
+
+## Sparse backend
+
+"""
+    AutoSparse{mode,D,S,C}
+
+Wraps an ADTypes.jl object to deal with sparse Jacobians and Hessians.
+
+# Fields
+
+- `dense_ad::D`: the underlying AD package, subtyping [`AbstractADType`](@ref)
+- `sparsity_detector::S`: the sparsity pattern detector, subtyping [`AbstractSparsityDetector`](@ref)
+- `coloring_algorithm::C`: the coloring algorithm, subtyping [`AbstractColoringAlgorithm`](@ref)
+
+# Constructors
+
+    AutoSparse(
+        dense_ad;
+        sparsity_detector=ADTypes.NoSparsityDetector(),
+        coloring_algorithm=ADTypes.NoColoringAlgorithm()
+    )
+"""
+struct AutoSparse{
+    mode, D <: AbstractADType{mode}, S <: AbstractSparsityDetector,
+    C <: AbstractColoringAlgorithm} <:
+       AbstractADType{mode}
+    dense_ad::D
+    sparsity_detector::S
+    coloring_algorithm::C
+end
+
+function AutoSparse(
+        dense_ad::AbstractADType{mode}; sparsity_detector = NoSparsityDetector(),
+        coloring_algorithm = NoColoringAlgorithm()) where {mode}
+    return AutoSparse{
+        mode, typeof(dense_ad), typeof(sparsity_detector), typeof(coloring_algorithm)}(
+        dense_ad, sparsity_detector, coloring_algorithm)
+end
+
+"""
+    dense_ad(ad::AutoSparse)::AbstractADType
+
+Return the underlying AD package for a sparse AD choice.
+    
+# See also
+
+- [`AutoSparse`](@ref)
+"""
+dense_ad(ad::AutoSparse) = ad.dense_ad
+
+"""
+    sparsity_detector(ad::AutoSparse)::AbstractSparsityDetector
+
+Return the sparsity pattern detector for a sparse AD choice.
+
+# See also
+
+- [`AutoSparse`](@ref)
+- [`AbstractSparsityDetector`](@ref)
+"""
+sparsity_detector(ad::AutoSparse) = ad.sparsity_detector
+
+"""
+    coloring_algorithm(ad::AutoSparse)::AbstractColoringAlgorithm
+
+Return the coloring algorithm for a sparse AD choice.
+
+# See also
+
+- [`AutoSparse`](@ref)
+- [`AbstractColoringAlgorithm`](@ref)
+"""
+coloring_algorithm(ad::AutoSparse) = ad.coloring_algorithm
+
+# TODO: document or remove these?
```

Comment:
Can probably just remove. Parametric dispatching seems like it should be used to handle this, or trait functions.

Reply:
My aim in keeping them was to be as gently breaking as possible, but if you greenlight it I can remove them.
Similar question for the deprecation

Reply:
Go for it